File: backend/internal/domain/waste_bank.go
package domain

import (
	"context"
	"time"

	"github.com/google/uuid"
)

type WasteCategory struct {
	ID                   uuid.UUID `json:"id"`
	TenantID             uuid.UUID `json:"tenant_id"`
	Name                 string    `json:"name"`
	Unit                 string    `json:"unit"` // kg, liter, pcs
	PricePerUnit         float64   `json:"price_per_unit"`
	ResidentSharePct     float64   `json:"resident_share_pct"`     // e.g. 80.0
	KarangTarunaSharePct float64   `json:"karang_taruna_share_pct"` // e.g. 20.0
	Description          *string   `json:"description,omitempty"`
	IsActive             bool      `json:"is_active"`
	CreatedAt            time.Time `json:"created_at"`
	UpdatedAt            time.Time `json:"updated_at"`
}

type WasteDepositItem struct {
	ID                 uuid.UUID `json:"id"`
	DepositID          uuid.UUID `json:"deposit_id"`
	CategoryID         uuid.UUID `json:"category_id"`
	CategoryName       string    `json:"category_name,omitempty"`
	CategoryUnit       string    `json:"category_unit,omitempty"`
	Quantity           float64   `json:"quantity"`
	UnitPrice          float64   `json:"unit_price"`
	GrossAmount        float64   `json:"gross_amount"`
	ResidentAmount     float64   `json:"resident_amount"`
	KarangTarunaAmount float64   `json:"karang_taruna_amount"`
	CreatedAt          time.Time `json:"created_at"`
}

type WasteDeposit struct {
	ID                 uuid.UUID           `json:"id"`
	TenantID           uuid.UUID           `json:"tenant_id"`
	HouseID            *uuid.UUID          `json:"house_id,omitempty"`
	ResidentID         *uuid.UUID          `json:"resident_id,omitempty"`
	KKNumber           *string             `json:"kk_number,omitempty"`
	FamilyHeadName     string              `json:"family_head_name"`
	DepositDate        time.Time           `json:"deposit_date"`
	TotalWeight        float64             `json:"total_weight"`
	TotalGrossAmount   float64             `json:"total_gross_amount"`
	ResidentAmount     float64             `json:"resident_amount"`
	KarangTarunaAmount float64             `json:"karang_taruna_amount"`
	Status             string              `json:"status"` // pending, verified, paid_out, cancelled
	Notes              *string             `json:"notes,omitempty"`
	RecordedBy         *uuid.UUID          `json:"recorded_by,omitempty"`
	Items              []*WasteDepositItem `json:"items,omitempty"`
	CreatedAt          time.Time           `json:"created_at"`
	UpdatedAt          time.Time           `json:"updated_at"`
}

type WasteBankSummary struct {
	TotalDeposits          int64   `json:"total_deposits"`
	TotalWeightKg          float64 `json:"total_weight_kg"`
	TotalGrossValue        float64 `json:"total_gross_value"`
	TotalResidentEarnings  float64 `json:"total_resident_earnings"`
	TotalKarangTarunaShare float64 `json:"total_karang_taruna_share"`
	ActiveHouseholdsCount  int64   `json:"active_households_count"`
}

type HouseholdWasteAccumulation struct {
	FamilyHeadName     string  `json:"family_head_name"`
	KKNumber           string  `json:"kk_number,omitempty"`
	TotalWeight        float64 `json:"total_weight"`
	TotalGrossAmount   float64 `json:"total_gross_amount"`
	ResidentAmount     float64 `json:"resident_amount"`
	KarangTarunaAmount float64 `json:"karang_taruna_amount"`
	DepositCount       int64   `json:"deposit_count"`
	LastDepositDate    string  `json:"last_deposit_date"`
}

type WasteBankRepository interface {
	// Categories
	ListCategories(ctx context.Context, tenantID uuid.UUID, activeOnly bool) ([]*WasteCategory, error)
	GetCategoryByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteCategory, error)
	CreateCategory(ctx context.Context, category *WasteCategory) error
	UpdateCategory(ctx context.Context, category *WasteCategory) error
	DeleteCategory(ctx context.Context, tenantID, id uuid.UUID) error

	// Deposits
	CreateDeposit(ctx context.Context, deposit *WasteDeposit) error
	GetDepositByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteDeposit, error)
	ListDeposits(ctx context.Context, tenantID uuid.UUID, search, status string, limit, offset int) ([]*WasteDeposit, int64, error)
	UpdateDepositStatus(ctx context.Context, tenantID, id uuid.UUID, status string) error

	// Aggregates
	GetSummary(ctx context.Context, tenantID uuid.UUID) (*WasteBankSummary, error)
	ListHouseholdAccumulations(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]*HouseholdWasteAccumulation, int64, error)
}

type WasteBankUsecase interface {
	ListCategories(ctx context.Context, tenantID uuid.UUID, activeOnly bool) ([]*WasteCategory, error)
	GetCategoryByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteCategory, error)
	CreateCategory(ctx context.Context, category *WasteCategory) error
	UpdateCategory(ctx context.Context, category *WasteCategory) error
	DeleteCategory(ctx context.Context, tenantID, id uuid.UUID) error

	CreateDeposit(ctx context.Context, deposit *WasteDeposit) error
	GetDepositByID(ctx context.Context, tenantID, id uuid.UUID) (*WasteDeposit, error)
	ListDeposits(ctx context.Context, tenantID uuid.UUID, search, status string, limit, offset int) ([]*WasteDeposit, int64, error)
	UpdateDepositStatus(ctx context.Context, tenantID, id uuid.UUID, status string) error

	GetSummary(ctx context.Context, tenantID uuid.UUID) (*WasteBankSummary, error)
	ListHouseholdAccumulations(ctx context.Context, tenantID uuid.UUID, limit, offset int) ([]*HouseholdWasteAccumulation, int64, error)
}

File: frontend/src/services/wasteBank.ts
import { useQuery } from '@tanstack/react-query';
import { api } from './api';

export interface WasteCategory {
  id: string;
  tenant_id: string;
  name: string;
  unit: string;
  price_per_unit: number;
  resident_share_pct: number;
  karang_taruna_share_pct: number;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface WasteDepositItem {
  id?: string;
  deposit_id?: string;
  category_id: string;
  category_name?: string;
  unit?: string;
  quantity: number;
  price_per_unit: number;
  subtotal: number;
  resident_share_amount: number;
  karang_taruna_share_amount: number;
}

export interface WasteDeposit {
  id: string;
  tenant_id: string;
  family_head_name: string;
  kk_number: string;
  rt_number?: string;
  house_number?: string;
  deposit_date: string;
  total_weight_kg: number;
  total_gross_amount: number;
  resident_earnings_amount: number;
  karang_taruna_amount: number;
  status: 'pending' | 'verified' | 'paid_out' | 'cancelled';
  payout_method: string;
  notes?: string;
  recorded_by?: string;
  created_at: string;
  updated_at: string;
  items?: WasteDepositItem[];
}

export interface WasteBankSummary {
  total_deposits: number;
  total_weight_kg: number;
  total_gross_value: number;
  total_resident_earnings: number;
  total_karang_taruna_share: number;
  active_households_count: number;
}

export interface HouseholdAccumulation {
  family_head_name: string;
  kk_number: string;
  rt_number: string;
  house_number: string;
  deposit_count: number;
  total_weight_kg: number;
  total_earnings_amount: number;
  last_deposit_date: string;
}

export const wasteBankService = {
  // Public
  getPublicCategories: async (slug: string): Promise<WasteCategory[]> => {
    const res = await api.get<{ data: WasteCategory[] }>(`/t/${slug}/waste-bank/categories`);
    return res.data.data;
  },
  getPublicSummary: async (slug: string): Promise<WasteBankSummary> => {
    const res = await api.get<WasteBankSummary>(`/t/${slug}/waste-bank/summary`);
    return res.data;
  },

  // Admin / Protected
  getCategories: async (activeOnly = false): Promise<WasteCategory[]> => {
    const res = await api.get<{ data: WasteCategory[] }>(`/waste-bank/categories`, {
      params: { active_only: activeOnly ? 'true' : 'false' }
    });
    return res.data.data;
  },
  createCategory: async (data: Partial<WasteCategory>): Promise<WasteCategory> => {
    const res = await api.post<{ data: WasteCategory }>(`/waste-bank/categories`, data);
    return res.data.data;
  },
  updateCategory: async (id: string, data: Partial<WasteCategory>): Promise<WasteCategory> => {
    const res = await api.put<{ data: WasteCategory }>(`/waste-bank/categories/${id}`, data);
    return res.data.data;
  },
  deleteCategory: async (id: string): Promise<void> => {
    await api.delete(`/waste-bank/categories/${id}`);
  },

  getDeposits: async (params?: { search?: string; status?: string; page?: number; limit?: number }) => {
    const res = await api.get<{ data: WasteDeposit[]; total: number }>(`/waste-bank/deposits`, { params });
    return res.data;
  },
  getDepositById: async (id: string): Promise<WasteDeposit> => {
    const res = await api.get<WasteDeposit>(`/waste-bank/deposits/${id}`);
    return res.data;
  },
  createDeposit: async (data: Partial<WasteDeposit>): Promise<WasteDeposit> => {
    const res = await api.post<{ data: WasteDeposit }>(`/waste-bank/deposits`, data);
    return res.data.data;
  },
  updateDepositStatus: async (id: string, status: string): Promise<void> => {
    await api.patch(`/waste-bank/deposits/${id}/status`, { status });
  },
  getSummary: async (): Promise<WasteBankSummary> => {
    const res = await api.get<WasteBankSummary>(`/waste-bank/summary`);
    return res.data;
  },
  getHouseholdAccumulations: async (page = 1, limit = 50) => {
    const res = await api.get<{ data: HouseholdAccumulation[]; total: number }>(`/waste-bank/households`, {
      params: { page, limit }
    });
    return res.data;
  },
};

export function useHouseholdAccumulationsQuery(page = 1, limit = 50) {
  return useQuery({
    queryKey: ['waste-bank-households', page, limit],
    queryFn: () => wasteBankService.getHouseholdAccumulations(page, limit),
  });
}
